web3.go

```diff
@@ -160,23 +164,64 @@ func CallTransactFunction(ctx context.Context, client Client, myabi abi.ABI, add
 	}
 	return convertTx(signedTx, fromAddress), nil
 }
+func isValidUrl(toTest string) bool {
+	_, err := url.ParseRequestURI(toTest)
+	if err != nil {
+		return false
+	}
+	u, err := url.Parse(toTest)
+	if err != nil || u.Scheme == "" || u.Host == "" {
+		return false
+	}
+	return true
+}
+func downloadFile(url string) ([]byte, error) {
+	var dst bytes.Buffer
+	response, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+	_, err = io.Copy(&dst, response.Body)
+	if err != nil {
+		return nil, err
+	}
+	return dst.Bytes(), nil
+}
 
 // DeployBin will deploy a bin file to the network
 func DeployBin(ctx context.Context, client Client,
 	privateKeyHex, binFilename, abiFilename string, gasPrice *big.Int, gasLimit uint64, constructorArgs ...interface{}) (*Transaction, error) {
-	bin, err := ioutil.ReadFile(binFilename)
-	if err != nil {
-		return nil, fmt.Errorf("Cannot read the bin file %q: %v", binFilename, err)
-	}
-	var abi string
-	if len(constructorArgs) > 0 {
-		b, err := ioutil.ReadFile(abiFilename)
+	var bin []byte
+	var err error
+	if isValidUrl(binFilename) {
+		bin, err = downloadFile(binFilename)
+		if err != nil {
+			return nil, fmt.Errorf("Cannot download the bin file %q: %v", binFilename, err)
+		}
+	} else {
+		bin, err = ioutil.ReadFile(binFilename)
 		if err != nil {
-			return nil, fmt.Errorf("Cannot read the abi file %q: %v", abiFilename, err)
+			return nil, fmt.Errorf("Cannot read the bin file %q: %v", binFilename, err)
+		}
+	}
+	var abi []byte
+	if isValidUrl(abiFilename) {
+		if len(constructorArgs) > 0 {
+			abi, err = downloadFile(abiFilename)
+			if err != nil {
+				return nil, fmt.Errorf("Cannot download the abi file %q: %v", abiFilename, err)
+			}
+		}
+	} else {
+		if len(constructorArgs) > 0 {
```

Comment:
nbd but seems like it would make sense to do the `if len(constructorArgs) > 0` check before doing `isValidUrl`, since it's cheaper and combines two duplicate lines.

Reply:
yep